File: LC/rough/local_practice/p43.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int ctr = 0;

    int countComponents(int n, vector<vector<int>> &edges)
    {
        vector<vector<int>> adj(n);
        vector<bool> visited(n, false);

        for (auto &vec : edges)
        {
            adj[vec[0]].push_back(vec[1]);
            adj[vec[1]].push_back(vec[0]);
        }

        for (int i = 0; i < n; i++)
        {
            if (!visited[i])
            {
                ctr++;
                dfs(i, adj, visited);
            }
        }

        return ctr;
    }

    void dfs(int node, vector<vector<int>> &adj, vector<bool> &visited)
    {
        visited[node] = true;

        for (int neigh : adj[node])
        {
            if (!visited[neigh])
            {
                dfs(neigh, adj, visited);
            }
        }
    }
};

File: LC/rough/local_practice/p43.py
# Definition for a Node.
class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


from typing import Optional


class Solution:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        clones = dict()
        if node is None:
            return

        def dfs(node):
            if node in clones:
                return

            new_node = Node(node.val)
            clones[node] = new_node

            for neigh in node.neighbors:
                dfs(neigh)
                new_node.neighbors.append(clones[neigh])

        dfs(node)
        return clones[node]
